lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,629 @@
+            end
+        end
+    end
+    if not grid["get_" .. prop] then
+        grid["get_"..prop] = function(self)
+            return self._private[prop]
+        end
+    end
+end
+
+--- Set the directional grid properties
+-- create a couple of properties by prepending `horizontal_` and `vertical_`
+-- create a common property for the two directions:
+-- setting the common property sets both directional properties
+-- getting the common property returns the directional property defined by the `orientation` property
+for _, prop in ipairs(dir_properties) do
+    for _,dir in ipairs{"horizontal_, vertical_"} do
+        local dir_prop = dir .. "_" .. prop
+        grid["set_"..dir_prop] = function(self, value)
+            if self._private[dir_prop] ~= value then
+                self._private[dir_prop] = value
+                self:emit_signal("widget::layout_changed")
+            end
+        end
+        grid["get_"..dir_prop] = function(self)
+            return self._private[dir_prop]
+        end
+    end
+
+    -- Non-directional options
+    grid["set_"..prop] = function(self, value)
+        if self._private["horizontal_"..prop] ~= value or self._private["vertical_"..prop] ~= value then
+            self._private["horizontal_"..prop] = value
+            self._private["vertical_"..prop] = value
+            self:emit_signal("widget::layout_changed")
+        end
+    end
+    grid["get_"..prop] = function(self)
+        if self._private.orientation == "x" then
+            return self._private["horizontal_"..prop]
+        elseif self._private.orientation == "y" then
+            return self._private["vertical_"..prop]
+        end
+    end
+end
+
+
+-- Return two tables of the fitted sizes of the rows and columns
+-- @return table,table Tables of row heights and column widths
+local function get_matrix_sizes(self, context, orig_width, orig_height)
+    local rows_size = {}
+    local cols_size = {}
+
+    -- Set the row and column sizes to the minimum value
+    for i = 1,self._private.num_rows do
+        rows_size[i] = self._private.min_rows_size
+    end
+    for j = 1,self._private.num_cols do
+        cols_size[j] = self._private.min_cols_size
+    end
+
+    -- Calculate cell sizes
+    for _, data in ipairs(self._private.widgets) do
+        local w, h = base.fit_widget(self, context, data.widget, orig_width, orig_height)
+        h = math.max( self._private.min_rows_size, h / data.row_span )
+        w = math.max( self._private.min_cols_size, w / data.col_span )
+        -- update the row and column maximum size
+        for i = data.row, data.row + data.row_span - 1 do
+            if h > rows_size[i] then rows_size[i] = h end
+        end
+        for j = data.col, data.col + data.col_span - 1 do
+            if w > cols_size[j] then cols_size[j] = w end
+        end
+    end
+    return rows_size, cols_size
+end
+
+
+--- Fit the grid layout into the given space
```

Comment:
Please use `--` for fit/draw/draw_*/layout as they are not "user facing" methods and don't need to be documented in the public doc over and over again.